server/src/main/java/org/apache/kafka/server/DelayedProduce.java

```diff
@@ -0,0 +1,235 @@
+        }
+    }
+
+    private final ProduceMetadata produceMetadata;
+    private final BiConsumer<TopicPartition, ProducePartitionStatus> updateProducePartitionStatusCallback;
+    private final Consumer<Map<TopicIdPartition, PartitionResponse>> responseCallback;
+
+    public DelayedProduce(long delayMs,
+                          ProduceMetadata produceMetadata,
+                          BiConsumer<TopicPartition, ProducePartitionStatus> updateProducePartitionStatusCallback,
+                          Consumer<Map<TopicIdPartition, PartitionResponse>> responseCallback) {
+        super(delayMs);
+
+        this.produceMetadata = produceMetadata;
+        this.updateProducePartitionStatusCallback = updateProducePartitionStatusCallback;
+        this.responseCallback = responseCallback;
+
+        // first update the acks pending variable according to the error code
+        produceMetadata.produceStatus.forEach((topicPartition, status) -> {
+            if (status.responseStatus.error == Errors.NONE) {
+                // Timeout error state will be cleared when required acks are received
+                status.acksPending = true;
+                status.responseStatus.error = Errors.REQUEST_TIMED_OUT;
+            } else {
+                status.acksPending = false;
+            }
+
+            LOGGER.trace("Initial partition status for {} is {}", topicPartition, status);
+        });
+    }
+
+    /**
+     * The delayed produce operation can be completed if every partition
+     * it produces to is satisfied by one of the following:
+     *
+     * Case A: Replica not assigned to partition
+     * Case B: Replica is no longer the leader of this partition
+     * Case C: This broker is the leader:
+     *   C.1 - If there was a local error thrown while checking if at least requiredAcks
+     *         replicas have caught up to this operation: set an error in response
+     *   C.2 - Otherwise, set the response with no error.
+     */
+    @Override
+    public boolean tryComplete() {
+        // check for each partition if it still has pending acks
+        produceMetadata.produceStatus.forEach((topicIdPartition, status) -> {
+            LOGGER.trace("Checking produce satisfaction for {}, current status {}", topicIdPartition, status);
+            // skip those partitions that have already been satisfied
+            if (status.acksPending) {
+                updateProducePartitionStatusCallback.accept(topicIdPartition.topicPartition(), status);
+            }
+        });
+
+        // check if every partition has satisfied at least one of case A, B or C
+        boolean anyPending = produceMetadata.produceStatus
+                                            .values()
+                                            .stream()
+                                            .anyMatch(ProducePartitionStatus::acksPending);
+        if (!anyPending) {
+            return forceComplete();
+        }
+
+        return false;
```

Comment:
nit:
```suggestion
        return produceMetadata.produceStatus.values()
            .stream()
            .findAny()
            .map(__ -> false)
            .orElseGet(this::forceComplete);
```

Reply:
hmm, I think the it keeps the idea of original implementation and easier to read. What do you think?

Reply:
I'm okay with it.